From 0155d3c7e2168d7376bb2f3ed62563f002ac21aa Mon Sep 17 00:00:00 2001
From: Lionel Landwerlin <lionel.g.landwerlin@intel.com>
Date: Mon, 17 Nov 2025 09:52:53 +0200
Subject: [PATCH] brw: fix workaround fence rlen field

send.ugm (1|M0)          r125     r0      null:0  0x0            0x0200651F           {$9} // wr:1+0, rd:0; fence invalid flush type scoped to tile

When destination of Send(s) is not null, the response length must not be 0.

Should only affect DG2 products.

Signed-off-by: Lionel Landwerlin <lionel.g.landwerlin@intel.com>
Cc: mesa-stable
Reviewed-by: Alyssa Rosenzweig <alyssa.rosenzweig@intel.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38478>
(cherry picked from commit 4816318887d63dddca16480cd953e336662324ad)
---
 .pick_status.json                         | 2 +-
 src/intel/compiler/brw/brw_workaround.cpp | 6 ++++--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/.pick_status.json b/.pick_status.json
index 6dad2bfe00306..0a8ebd52388d8 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -1144,7 +1144,7 @@
         "description": "brw: fix workaround fence rlen field",
         "nominated": true,
         "nomination_type": 1,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": null,
         "notes": null
diff --git a/src/intel/compiler/brw/brw_workaround.cpp b/src/intel/compiler/brw/brw_workaround.cpp
index 7dee3c966745e..44106915935cd 100644
--- a/src/intel/compiler/brw/brw_workaround.cpp
+++ b/src/intel/compiler/brw/brw_workaround.cpp
@@ -120,9 +120,11 @@ brw_workaround_memory_fence_before_eot(brw_shader &s)
       dummy_fence->mlen = reg_unit(s.devinfo);
       dummy_fence->ex_mlen = 0;
       dummy_fence->sfid = BRW_SFID_UGM;
-      dummy_fence->desc = lsc_fence_msg_desc(s.devinfo, LSC_FENCE_TILE,
-                                             LSC_FLUSH_TYPE_NONE_6, false);
       dummy_fence->size_written = REG_SIZE * reg_unit(s.devinfo);
+      dummy_fence->desc = lsc_fence_msg_desc(s.devinfo, LSC_FENCE_TILE,
+                                             LSC_FLUSH_TYPE_NONE_6, false) |
+                          brw_message_desc(s.devinfo, dummy_fence->mlen,
+                                           dummy_fence->size_written / REG_SIZE, 0);
       ubld.emit(FS_OPCODE_SCHEDULING_FENCE, ubld.null_reg_ud(), dst);
       progress = true;
       /* TODO: remove this break if we ever have shader with multiple EOT. */
-- 
GitLab

